ILSpy/AssemblyTree/AssemblyTreeModel.cs

```diff
@@ -0,0 +1,932 @@
+						DecompileSelectedNodes(
+							DockWorkspace.Instance.ActiveTabPage.GetState() as DecompilerTextViewState);
+						break;
+					case nameof(SessionSettings.CurrentCulture):
+						MessageBox.Show(Properties.Resources.SettingsChangeRestartRequired, "ILSpy");
+						break;
+				}
+			}
+			else if (sender is LanguageSettings)
+			{
+				switch (e.PropertyName)
+				{
+					case nameof(LanguageSettings.Language) or nameof(LanguageSettings.LanguageVersion):
+						DecompileSelectedNodes(recordHistory: false);
+						break;
+				}
+			}
+		}
+
+		public AssemblyList AssemblyList { get; private set; }
+
+		private SharpTreeNode root;
+		public SharpTreeNode Root {
+			get => root;
+			set => SetProperty(ref root, value);
+		}
+
+		private SharpTreeNode selectedItem;
+		public SharpTreeNode SelectedItem {
+			get => selectedItem;
+			set => SetProperty(ref selectedItem, value);
+		}
+
+		public ObservableCollection<SharpTreeNode> SelectedItems { get; } = [];
+
+		public string[] SelectedPath => GetPathForNode(SelectedItem);
+
+		readonly List<LoadedAssembly> commandLineLoadedAssemblies = [];
+
+		public bool HandleCommandLineArguments(CommandLineArguments args)
+		{
+			LoadAssemblies(args.AssembliesToLoad, commandLineLoadedAssemblies, focusNode: false);
+			if (args.Language != null)
+				SettingsService.Instance.SessionSettings.LanguageSettings.Language = Languages.GetLanguage(args.Language);
+			return true;
+		}
+
+		/// <summary>
+		/// Called on startup or when passed arguments via WndProc from a second instance.
+		/// In the format case, spySettings is non-null; in the latter it is null.
+		/// </summary>
+		public void HandleCommandLineArgumentsAfterShowList(CommandLineArguments args, ISettingsProvider spySettings = null)
+		{
+			var sessionSettings = SettingsService.Instance.SessionSettings;
+
+			var relevantAssemblies = commandLineLoadedAssemblies.ToList();
+			commandLineLoadedAssemblies.Clear(); // clear references once we don't need them anymore
+			NavigateOnLaunch(args.NavigateTo, sessionSettings.ActiveTreeViewPath, spySettings, relevantAssemblies);
+			if (args.Search != null)
+			{
+				var searchPane = App.ExportProvider.GetExportedValue<SearchPaneModel>();
+
+				searchPane.SearchTerm = args.Search;
+				searchPane.Show();
+			}
+		}
+
+		public async Task HandleSingleInstanceCommandLineArguments(string[] args)
+		{
+			var cmdArgs = CommandLineArguments.Create(args);
+
+			await Dispatcher.InvokeAsync(() => {
+
+				if (!HandleCommandLineArguments(cmdArgs))
+					return;
+
+				var window = Application.Current.MainWindow;
+
+				if (!cmdArgs.NoActivate && window is { WindowState: WindowState.Minimized })
+				{
+					window.WindowState = WindowState.Normal;
+				}
+
+				HandleCommandLineArgumentsAfterShowList(cmdArgs);
+			});
+		}
+
+		public async void NavigateOnLaunch(string navigateTo, string[] activeTreeViewPath, ISettingsProvider spySettings, List<LoadedAssembly> relevantAssemblies)
```

Comment:
Please don't use `async void` unless it's absolutely necessary. And, if it's necessary call-sites need a comment hinting at the fact and we need to make sure it does not cause problems.

Reply:
Fully agree, however this is not new code, it is just moved from `MainWindow.xaml.cs` an I did not want to introduce too much functional changes in this PR, but just decouple UI and model, so maybe we can go with this and clean it up later since this PR is already huge.